File: src/main/java/no/fintlabs/flyt/gateway/application/archive/template/model/ObjectTemplate.java
package no.fintlabs.flyt.gateway.application.archive.template.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import no.fintlabs.flyt.gateway.application.archive.template.validation.UniqueKeys;
import no.fintlabs.flyt.gateway.application.archive.template.validation.UniqueOrders;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;

@UniqueKeys
@UniqueOrders
@Getter
@EqualsAndHashCode
@JsonInclude(JsonInclude.Include.NON_EMPTY)
public class ObjectTemplate {

    public static ObjectTemplateBuilder builder() {
        return new ObjectTemplateBuilder();
    }

    @JsonCreator
    public ObjectTemplate(
            @JsonProperty("valueTemplates") Collection<ElementTemplate<ValueTemplate>> valueTemplates,
            @JsonProperty("selectableValueTemplates") Collection<ElementTemplate<SelectableValueTemplate>> selectableValueTemplates,
            @JsonProperty("valueCollectionTemplates") Collection<ElementTemplate<CollectionTemplate<ValueTemplate>>> valueCollectionTemplates,
            @JsonProperty("objectTemplates") Collection<ElementTemplate<ObjectTemplate>> objectTemplates,
            @JsonProperty("objectCollectionTemplates") Collection<ElementTemplate<CollectionTemplate<ObjectTemplate>>> objectCollectionTemplates
    ) {
        this.valueTemplates = Optional.ofNullable(valueTemplates).orElse(new ArrayList<>());
        this.selectableValueTemplates = Optional.ofNullable(selectableValueTemplates).orElse(new ArrayList<>());
        this.valueCollectionTemplates = Optional.ofNullable(valueCollectionTemplates).orElse(new ArrayList<>());
        this.objectTemplates = Optional.ofNullable(objectTemplates).orElse(new ArrayList<>());
        this.objectCollectionTemplates = Optional.ofNullable(objectCollectionTemplates).orElse(new ArrayList<>());
    }

    private final Collection<@NotNull @Valid ElementTemplate<ValueTemplate>> valueTemplates;
    private final Collection<@NotNull @Valid ElementTemplate<SelectableValueTemplate>> selectableValueTemplates;
    private final Collection<@NotNull @Valid ElementTemplate<CollectionTemplate<ValueTemplate>>> valueCollectionTemplates;
    private final Collection<@NotNull @Valid ElementTemplate<ObjectTemplate>> objectTemplates;
    private final Collection<@NotNull @Valid ElementTemplate<CollectionTemplate<ObjectTemplate>>> objectCollectionTemplates;

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/template/model/ObjectTemplate.kt
package no.novari.flyt.archive.gateway.template.model

import com.fasterxml.jackson.annotation.JsonInclude
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import no.novari.flyt.archive.gateway.template.validation.UniqueKeys
import no.novari.flyt.archive.gateway.template.validation.UniqueOrders

@UniqueKeys
@UniqueOrders
@JsonInclude(JsonInclude.Include.NON_EMPTY)
data class ObjectTemplate(
    val valueTemplates: Collection<
        @NotNull @Valid
        ElementTemplate<ValueTemplate>,
    > = emptyList(),
    val selectableValueTemplates: Collection<
        @NotNull @Valid
        ElementTemplate<SelectableValueTemplate>,
    > = emptyList(),
    val valueCollectionTemplates: Collection<
        @NotNull @Valid
        ElementTemplate<CollectionTemplate<ValueTemplate>>,
    > = emptyList(),
    val objectTemplates: Collection<
        @NotNull @Valid
        ElementTemplate<ObjectTemplate>,
    > = emptyList(),
    val objectCollectionTemplates: Collection<
        @NotNull @Valid
        ElementTemplate<CollectionTemplate<ObjectTemplate>>,
    > = emptyList(),
) {
    companion object {
        @JvmStatic
        fun builder() = ObjectTemplateBuilder()
    }
}
